### PR TITLE
Add Shim for wakeup_fds with no pipe or 

### PR DESCRIPTION
**Pull request by kpayson64**
Nacl does not support pipe.  It looks like they might support eventfd if __NR_eventfd is set, but this is not the case for all builds.

This is a shim that uses pthread conditions to simulate event fds.

### CODE REVIEW

File: include/grpc/impl/codegen/port_platform.h [hunk 1]
```diff
@@ -196,7 +196,7 @@
 #define GPR_GCC_TLS 1
 #define GPR_LINUX 1
 #define GPR_LINUX_LOG
-#define GPR_LINUX_MULTIPOLL_WITH_EPOLL 1
+#define GPR_POSIX_MULTIPOLL_WITH_POLL 1
```
Comment (murgatroid99) [inline include/grpc/impl/codegen/port_platform.h]: I think this changes a constant that is currently used to one that is not currently used.

Also, why are you changing the `__linux__` section of `port_platform.h` for a change related to NACL?

Comment (nicolasnoble) [inline include/grpc/impl/codegen/port_platform.h]: Removing multipoll with epoll from default Linux will cause performance degradation :)

Comment (kpayson64) [inline include/grpc/impl/codegen/port_platform.h]: I intend to remove this before submission.  I wanted to run all the linux c/wrapped language tests using this shim to be more confident that it works.  

Eventually I want to add a whole NaCl build column to Jenkins and run the c tests using a nacl build, but that is outside the scope of this PR.

This was a good review catch btw :)

Comment (murgatroid99) [inline include/grpc/impl/codegen/port_platform.h]: It happens that I'm currently working on my own change that touches the same part of port_platform.h

Comment (ctiller) [inline include/grpc/impl/codegen/port_platform.h]: For testing you can use the environment variable 'GRPC_POLL_STRATEGY=poll' to force disabling epoll.

File: src/core/lib/support/thd_posix.c [hunk 2]
```diff
@@ -92,4 +92,6 @@ gpr_thd_id gpr_thd_currentid(void) { return (gpr_thd_id)pthread_self(); }
 
 void gpr_thd_join(gpr_thd_id t) { pthread_join((pthread_t)t, NULL); }
 
+void gpr_thd_cancel(gpr_thd_id t) { pthread_cancel((pthread_t)t); }
```
Comment (ctiller) [inline src/core/lib/support/thd_posix.c]: Please no. pthread_cancel considered harmful.

Comment (kpayson64) [inline src/core/lib/support/thd_posix.c]: It turned out ```pthread_cancel()``` was not available for the target platform I had in mind, so this is a vestige of a previous attempt.  I'll remove it.

**Comment by kpayson64** [comment]
Cleaned up the platform detection.

@ctiller PTAL

**Comment by kpayson64** [comment]
As it turns out, not all the pollers are tested by default by jenkins.  Here is a build where I forced the poller to be poll-cv:
https://grpc-testing.appspot.com/job/gRPC_pull_requests/13330/config=tsan,language=c,platform=linux/

@ctiller 
This is ready for review

**Comment by kpayson64** [comment]
My previous comment is no longer true with some jenkins changes that happened.

File: src/core/lib/iomgr/ev_poll_cv_posix.c [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+
+const grpc_event_engine_vtable* grpc_init_poll_cv_posix(void) {
+  int has_wakeup_fd = grpc_has_wakeup_fd;
+  int allow_specialized_wakeup_fd = grpc_allow_specialized_wakeup_fd;
+  int allow_pipe_wakeup_fd = grpc_allow_pipe_wakeup_fd;
+  grpc_global_cv_fd_table_init();
+  grpc_allow_specialized_wakeup_fd = 0;
+  grpc_allow_pipe_wakeup_fd = 0;
+  grpc_wakeup_fd_global_init();
+  grpc_has_wakeup_fd = 1;
```
Comment (ctiller) [inline src/core/lib/iomgr/ev_poll_cv_posix.c]: Let's not fiddle this externally owned variable from here.

Instead:
1. let's make grpc_has_wakeup_fd a function
2. move this code into ev_poll_posix.c
3. utilize the special (internal) knowledge that gives us to provide this implementation

Comment (kpayson64) [inline src/core/lib/iomgr/ev_poll_cv_posix.c]: I agree that this needs some cleanup.  If I understand your suggestion correctly, we won't be able to select "poll-cv" as our polling engine, which will make running exhaustive tests more difficult.

I've attempted to clean up this setup while leaving poll-cv as an independent poller.

Comment (kpayson64) [inline src/core/lib/iomgr/ev_poll_cv_posix.c]: Makes sense.  Fixed.
